File: bankomat/typescript/src/models/cash-machine.ts
import { BankProvider } from "../services/bank-provider.service"
import { Card } from "./card"

export class CashMachine {
    private bankProvider: BankProvider
    private stateOfMoney: number
    private inputedCard?: Card
    private inputedPin?: number
    private imputedAmount?: number
    private isLogged = false;
    private maintenanceLogingCode: number

    constructor (stateOfMoney:number, maintenanceCode:number){
        this.stateOfMoney = stateOfMoney
        this.bankProvider = BankProvider.getInstance()
        this.maintenanceLogingCode = maintenanceCode
    }


    public insertCard(card: Card){
        if(this.inputedCard){
            return console.log("Error")
        } else {
            this.inputedCard = card
        }
    }

    public insertPin(pin: number){
        this.inputedPin = pin
    }

    public logIn(){
        if (this.checkPinInBank()){
            console.log("you are succesfully logged")
            this.isLogged = true
        }
    }

    public logOut(){
        this.isLogged = false;
        this.inputedCard = undefined;
        this.inputedPin = undefined;
        console.log("you are succesfully logout")
    }

    public insertAmount(amount: number){
        if (this.isLogged && amount > 0){
            if(this.stateOfMoney >=amount ){
                this.imputedAmount = amount
                console.log(`cash to withrdaw: ${amount}`)
            }
            else console.log(`sorry atm have not enought cash to withdraw`)
        }
        else console.log(`you must first log in`)
    }

    public payOut(){
        if(this.isLogged){
            if (this.imputedAmount){
                if (this.payOutFromBank()) {
                    this.stateOfMoney -= this.imputedAmount
                    console.log(`You withdrew ${this.imputedAmount}`)
                    this.imputedAmount = undefined;
                } else {
                    console.log('DB ERROR')
                }
            } else {
                console.log(`Please input amount`)
            }

        }
        else console.log(`you must first log in`)
    }

    public payIn() {
        if(this.isLogged){
            if (this.imputedAmount){
                if (this.payInFromBank()) {
                    this.stateOfMoney += this.imputedAmount
                    console.log(`You deposit ${this.imputedAmount}`)
                    this.imputedAmount = undefined;
                } else {
                    console.log('DB ERROR')
                }
            } else {
                console.log(`Please input amount`)
            }
        }
        else console.log(`you must first log in`)
    }

    public maintenance(code: number, maintenance: number){
        if (this.maintenanceLogingCode === code){
            if(maintenance === 1){
                this.checkAtmStatus()
            }
            else if(maintenance === 2){
                console.log(`opening cashbox`)
            }
            else if(maintenance === 3){
                console.log(`restaring...`)
            }
            else console.log(`comend unrecognized`)

        }
        else console.log(`wrong maintenence code`)
    }

    public checkBalance(){
        if(this.isLogged){
            if(this.inputedCard?.id && this.inputedPin){

                const saldo = this.bankProvider.checkBalanceFromBank(this.inputedCard.id, this.inputedPin)

                if (saldo === null) {
                    console.log(`Unknown saldo in your bank account`)
                } else {
                    console.log(`You have ${saldo} money in your bank account`)
                }
            }
        } else {
            console.log(`you must first insert amout of money you want withdraw`)
        }
    }

    private checkAtmStatus(){
        return console.log(`money left in casbox ${this.stateOfMoney}`)
    }

    private checkPinInBank(){
        if(this.inputedCard?.id && this.inputedPin){
            return this.bankProvider.checkPin(this.inputedCard.id, this.inputedPin)
        } else {
            console.log(`you must first insert amout of money you want withdraw`)
        }
    }

    private payOutFromBank(){
        if(this.inputedCard?.id && this.inputedPin && this.imputedAmount){
            return this.bankProvider.payOut(this.inputedCard.id, this.inputedPin, this.imputedAmount)
        }
    }

    private payInFromBank() {
        if(this.inputedCard?.id && this.inputedPin && this.imputedAmount){
            return this.bankProvider.payIn(this.inputedCard.id, this.inputedPin, this.imputedAmount)
        }
    }
}

File: bankomat/python/models/CashMachine.py
from services.BankProvider import BankProvider

class CashMachine():
    def __init__(self, moneyAmount):

        self.__moneyAmount = moneyAmount
        self.__bankProvider = BankProvider()
        self.__inputedCard = None
        self.__inputedPin = None
        self.__inputedAmount = None
        self.__isLogged = False

    def insertCard(self, card):
        if self.__inputedCard != None:
            return "Error"

        self.__inputedCard = card

    def insertPin(self, pin):
        self.__inputedPin = pin

    def __checkPinInBank(self):
        if self.__inputedCard and self.__inputedCard.number and self.__inputedPin:
            return self.__bankProvider.checkPin(self.__inputedCard.number, self.__inputedPin)
        else:
            print("Error: you must first insert amout of money you want withdraw")

    def __payOutFromBank(self):
        if self.__inputedCard and self.__inputedCard.number and self.__inputedPin:
            return self.__bankProvider.payOut(self.__inputedCard.number, self.__inputedPin, self.__inputedAmount)

    def logIn(self):
        if self.__checkPinInBank():
            print("you are succesfully logged")
            self.isLogged = True
        else:
            print("Error: wrong pin or card number")
            self.isLogged = False

        return self.isLogged

    def insertAmount(self, amount):
        if self.isLogged:
            if amount <= self.__moneyAmount:
                self.__inputedAmount = amount
                print(f"cash to withdraw: {amount}")
            else:
                print("Error: sorry atm have not enought cash to withdraw")
        else:
            print("Error: you aren't logged in")

    def payOut(self):
        if self.isLogged:
            if self.__inputedAmount:
                if(self.__payOutFromBank()):
                    self.__moneyAmount -= self.__inputedAmount
                    print(f"you withdrew {self.__inputedAmount}")
                else:
                    print("Pay out from bank error")
            else:
                print("Please choose amount")

        else:
            print("you aren't logged in")

    def checkBalance(self):
        if self.isLogged:
            if self.__inputedCard.number and self.__inputedPin:
                balance = self.__bankProvider.checkBalanceFromBank(
                    self.__inputedCard.number, self.__inputedPin)
                if balance == None:
                    print("Unknown saldo in your bank account")
                else:
                    print(f"You have {balance} money in your bank account")
            else:
                print("Error: firstly you must insert amount of money you want to withdraw")
